pkg/controller/datavolume/import-controller.go

```diff
@@ -198,58 +204,60 @@ func (r ImportReconciler) updateAnnotations(dataVolume *cdiv1.DataVolume, annota
-	if syncErr != nil {
-		return reconcile.Result{}, syncErr
-	}
-	if syncRes.res != nil {
-		return *syncRes.res, nil
+// Reconcile loop for the import data volumes
+func (r ImportReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("DataVolume", req.NamespacedName)
+	return r.updateStatus(r.sync(log, req))
+}
+
+func (r ImportReconciler) sync(log logr.Logger, req reconcile.Request) (dataVolumeSyncResult, error) {
+	var syncRes dataVolumeSyncResult
+	syncErr := r.syncCommon(log, req, &syncRes, nil, nil)
+	if syncErr != nil || syncRes.result != nil || syncRes.gc {
+		return syncRes, syncErr
 	}
-	//FIXME: pass syncRes instead of args
-	if err := r.reconcileDataVolumePVC(log, syncRes.dv, &syncRes.pvc, syncRes.pvcSpec); err != nil {
-		return reconcile.Result{}, err
+	if syncRes.pvc == nil {
+		if _, dvPrePopulated := syncRes.dv.Annotations[cc.AnnPrePopulated]; !dvPrePopulated {
+			syncRes.pvc, syncErr = r.createPvcForDatavolume(syncRes.dv, syncRes.pvcSpec, r.updateAnnotations)
+		}
 	}
-	return r.reconcileDataVolumeStatus(syncRes.dv, syncRes.pvc, nil, r.updateStatusPhase)
+	if syncRes.pvc != nil && syncErr == nil {
+		syncErr = r.maybeSetPvcMultiStageAnnotation(syncRes.pvc, syncRes.dv)
 
+	}
+	return syncRes, syncErr
 }
 
-func (r *ImportReconciler) reconcileDataVolumePVC(log logr.Logger, dv *cdiv1.DataVolume, pvc **corev1.PersistentVolumeClaim, pvcSpec *v1.PersistentVolumeClaimSpec) error {
-	if _, dvPrePopulated := dv.Annotations[cc.AnnPrePopulated]; dvPrePopulated {
-		return nil
-	}
-	if *pvc == nil {
-		newPvc, err := r.createPvcForDatavolume(dv, pvcSpec, r.updateCheckpoint)
-		if err != nil {
-			if cc.ErrQuotaExceeded(err) {
-				r.updateDataVolumeStatusPhaseWithEvent(cdiv1.Pending, dv, nil, nil,
-					Event{
-						eventType: corev1.EventTypeWarning,
-						reason:    cc.ErrExceededQuota,
-						message:   err.Error(),
-					})
+func (r ImportReconciler) updateStatus(syncRes dataVolumeSyncResult, syncErr error) (reconcile.Result, error) {
+	if syncErr != nil {
+		if cc.ErrQuotaExceeded(syncErr) {
+			err := r.updateDataVolumeStatusPhaseWithEvent(cdiv1.Pending, syncRes.dv, nil, nil,
+				Event{
+					eventType: corev1.EventTypeWarning,
+					reason:    cc.ErrExceededQuota,
+					message:   syncErr.Error(),
+				})
+			if err != nil {
+				syncErr = err
 			}
-			return err
+			return getReconcileResult(syncRes.result), syncErr
 		}
-		*pvc = newPvc
-		return nil
 	}
-	if cc.GetSource(*pvc) == cc.SourceVDDK {
-		changed, err := r.getVddkAnnotations(dv, *pvc)
-		if err != nil {
-			return err
-		}
-		if changed {
-			err = r.client.Get(context.TODO(), types.NamespacedName{Name: dv.Name, Namespace: dv.Namespace}, dv)
-			if err != nil {
-				return err
+	modifier := func(dv *cdiv1.DataVolume) {
+		if syncRes.pvc != nil && cc.GetSource(syncRes.pvc) == cc.SourceVDDK {
+			if vddkHost := syncRes.pvc.Annotations[cc.AnnVddkHostConnection]; vddkHost != "" {
```

Comment:
One of the goals of this controller refactoring work should be to enable the status subresource on DataVolumes.   Not sure if you've thought much about that.  But it may be good to make sure that `updateStatus()` only updates the resource status.  In which case, updates like this (adding annotations) should be moved to `sync()`

Reply:
fixed

Reply:
Confirming your fix for now is to have the annotations added during `sync()` but still relying on `updateStatus()` to call update?

Reply:
Right, I guess we don't want to call `client.Update()` both on `sync()` and `updateStatus()`. You may also have both annotations and status updates on the same CR, so a single `client.Update()` will be called.

Reply:
I understand that fewer api calls makes sense for efficiency.  But I think it may be nice to stick to the theme of this pr and update the resources in sync now.  Then in `emitEvent()` you can change:

`if !reflect.DeepEqual(dataVolume, dataVolumeCopy) {`

to 

`if !reflect.DeepEqual(dataVolume.Status, dataVolumeCopy.Status) {`


also, I see that `updateStatusCommon()` may annotate the DataVolume like here:  https://github.com/arnongilboa/containerized-data-importer/blob/sync_update_status/pkg/controller/datavolume/controller.go#L464


Basically, I'm thinking that it may make sense to go over the sync/updateStatus code paths now.  For `sync() `call `update()` if the resource spec/metadata changes.  Make sure `sync()` does not update any status values.  for `updateStatus()`, make sure no spec/metadata changes occur.

Then in a follow up pr, you can enable the status subresource for the DataVolume.  Change `updateStatus()` to to call `Status().Update()` and see if anything breaks.

What do you think?

Reply:
Sounds great. Fixing.